### PR TITLE
Close MySQL connections in MetricBeat

### LINKED ISSUE #9415: MySQL Metricbeat module do not close db connection
The MySQL metricbeat module opens a connection to the database here https://github.com/elastic/beats/blob/master/metricbeat/module/mysql/mysql.go#L103

It seems that the `Close()` call is not being done as needed by the Golang `sql` driver https://golang.org/pkg/database/sql/driver/#Conn

### PR DESCRIPTION
**Pull request by cachedout**
This corrects an issue present in MetricBeat's MySQL module. Prior to this patch, MySQL connections were left half-open. The effects of this could be observed in the test suite, which would produce errors such as the following:

```
2018-12-08T02:19:47.548050Z 498 [Note] Aborted connection 498 to db: 'unconnected' user: 'root' host: '172.19.0.1' (Got an error reading communication packets)
2018-12-08T02:19:48.569906Z 500 [Note] Aborted connection 500 to db: 'unconnected' user: 'root' host: '172.19.0.1' (Got an error reading communication packets)
2018-12-08T02:19:48.570422Z 497 [Note] Aborted connection 497 to db: 'unconnected' user: 'root' host: '172.19.0.1' (Got an error reading communication packets)
2018-12-08T02:19:55.726372Z 508 [Note] Aborted connection 508 to db: 'unconnected' user: 'root' host: '172.19.0.1' (Got an error reading communication packets)
```

*Questions:* I chose a slightly different variable name for the error variable. Elsewhere in the codebase it is `err` everywhere so far as I can tell. Is there a preference by the developers to standardize on `err`? It can be easily changed. :) I chose the alternate name to clearly distinguish it from the variable of the same name earlier in the function.

Also, should issues like this be entered into the CHANGELOG? I did so just to be safe but it's so small that I'm not sure what the standard is here. 

Finally, I can't find any documentation that includes guidelines about which is the correct branch to commit to. Is this right?

Closes #9415

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -59,6 +59,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha1...master[Check the HEAD d
 - Fix panic on docker healthcheck collection on dockers without healthchecks. {pull}9171[9171]
 - Fix issue with not collecting Elasticsearch cross-cluster replication stats correctly. {pull}9179[9179]
 - The `node.name` field in the `elasticsearch/node` metricset now correctly reports the Elasticsarch node name. Previously this field was incorrectly reporting the node ID instead. {pull}9209[9209]
+- Fix issue where MySQL connections were not properly closed {issue}9415[9415]
```
Comment (jsoriano) [inline CHANGELOG.asciidoc]: Oh, and add also a reference to this pull request here please.

Comment (sayden) [reply CHANGELOG.asciidoc]: The reference, which is placed at the end of the CHANGELOG line; usually looks like `{pull}pull-number[pull-number]` You can see some examples in the same doc

**Comment by sayden** [comment]
jenkins, test this

**Comment by cachedout** [comment]
Whilst poking around here a bit more I discovered some odd (to me at least) behavior which I have detailed here: https://discuss.elastic.co/t/metricbeat-seems-to-be-reading-the-same-config-value-and-getting-different-values/160438

**Comment by cachedout** [comment]
I propose going with the first option. We could do a close on every iteration, but we'd have to modify the function to explicitly open it again as you mention since the db handle won't be set to nil after a close. The overhead of re-opening connections every time strikes me as unnecessary overhead, and of course, Beats should be as non-intrusive as possible, no? :)

Regarding the addition of a Close() ala what's in the Kubernetes module. I don't believe any _additional_ connection handling is necessary because this doesn't change the normal operation of the Fetch loop since the Close() should only ever be called on the termination of the module itself as I understand it. I've tested a database going away from underneath metricbeat while it is operating and it appears that the underlying db library is doing the right thing and handling reconnection on its own.

However. :)

It doesn't look to me like Close() function gets called at *all* when metricbeat shuts down. Perhaps I don't understand what you mean by "the metricset is stopped"? I assume this means that metricbeat itself is killed via term or kill signal but perhaps you mean something else? Happy to post some sample code if we want to test this further.

Thoughts?

**Comment by andrewkroh** [comment]
`Close()` should called on a clean shutdown or anytime the metricset is unloaded (in the case of `reload: true` being used with modules.d).

https://github.com/elastic/beats/blob/824d443bc182a12f30859667da4739beef9c8606/metricbeat/mb/module/wrapper.go#L132

That being said, I don’t see a test asserting Close() is called in the [unit tests](
https://github.com/elastic/beats/blob/824d443bc182a12f30859667da4739beef9c8606/metricbeat/mb/module/wrapper_test.go#L66).

**Comment by jsoriano** [comment]
`Close()` should be also called when modules are stopped on [autodiscover](https://www.elastic.co/guide/en/beats/metricbeat/6.5/configuration-autodiscover.html) configuration reload.

**Comment by ruflin** [comment]
@cachedout Any updates here? Would be great to get this fix in.

**Comment by cachedout** [comment]
@ruflin I can get to this by the end of the week, I think. Apologies for letting this go stale.

**Comment by elasticcla** [comment]
Hi @cachedout, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in your Git [commit](https://github.com/elastic/beats/pull/9449.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

File: metricbeat/module/mysql/status/status.go [hunk 2]
```diff
@@ -71,6 +71,12 @@ func (m *MetricSet) Fetch() (common.MapStr, error) {
 		return nil, err
 	}
 
+	var cerr error
+	cerr = m.db.Close()
```
[the remark this answers is not in the document]
Comment (cachedout) [reply metricbeat/module/mysql/status/status.go]: Changed to `defer`. Is that what you were looking for here?

**Comment by jsoriano** [comment]
This was fixed by #10355

**Pull request closed without merging**